src/main/java/edu/ucsb/cs56/pconrad/springboot/hello/DatabaseAPI.java

```diff
@@ -0,0 +1,342 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+// import java.util.concurrent.TimeUnit;
+import java.util.concurrent.CountDownLatch;
+import java.io.FileInputStream;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.FirebaseDatabase;
+import com.google.firebase.database.ValueEventListener;
+
+public class DatabaseAPI {
+    private static final String authFileName = "experimental-prj-firebase-adminsdk.json";
+
+    // initialize database access
+    static {
+        try {
+            FileInputStream serviceAccount =
+                    new FileInputStream(DatabaseAPI.authFileName);
+
+            FirebaseOptions options = new FirebaseOptions.Builder()
+                    .setCredentials(GoogleCredentials.fromStream(serviceAccount))
+                    .setDatabaseUrl("https://experimental-prj.firebaseio.com")
+                    .build();
+
+            FirebaseApp.initializeApp(options);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static final FirebaseDatabase database = FirebaseDatabase.getInstance();
+
+    // DONE
+    // createUser()
+    public static boolean createUser(User user) {
+        // check if the userid already exists
+        if (DatabaseAPI.findUser(user.getUserid()) != null) { return false; }
+        DatabaseReference ref = DatabaseAPI.database.getReference("users");
+        ref.child(user.getUserid()).setValueAsync(user);
+        return true;
+    }
+
+    // DONE
+    // findUser()
+    public static User findUser(String userid) {
+        DatabaseReference ref = DatabaseAPI.database.getReference("users").child(userid);
+        CountDownLatch doneSignal = new CountDownLatch(1);
+        List<User> temp = new ArrayList<>();
+
+        // System.out.println("Before read");
+        ref.addListenerForSingleValueEvent(new ValueEventListener() {
+            @Override
+            public void onDataChange(DataSnapshot dataSnapshot) {
+                // System.out.println(dataSnapshot);
+                User target = dataSnapshot.getValue(User.class);
+                if (target != null)
+                    temp.add(target);
+                doneSignal.countDown();
+            }
+            @Override
+            public void onCancelled(DatabaseError databaseError) {
+                System.out.println("The read failed: " + databaseError.getCode());
+                doneSignal.countDown();
+            }
+        });
+        // System.out.println("End read");
+        // System.out.println("Begin wait");
+        try {
+            doneSignal.await();
```

Comment:
While in general, this await should always return, in the event of some network error or some change to the underlying library, it may be possible for this to never return if onDataChange or onCancelled is never called.

I'd suggest you add a long timeout to this to make sure that even if there are other issues, your code will never be stuck here. You can use something reasonable such as 60 seconds, which under 99.9% of cases should never take 60 seconds for the data to load.

Reply:
Resolved.